### ISSUE TITLE
[Bug]: Gated DeltaNet Recurrence with constant Gate at 1 vs. DeltaNet Recurrence

### ISSUE THREAD

**Issue by JulienSiems** [fla-org/flash-linear-attention#104]
### Describe the bug

Hey everyone, 

I was comparing the output of the gated deltanet recurrence with the deltanet recurrence (both chunk wise form) with the reference implementation. In my case the gate needs to be 1, so I set it to zero (because it's in log space) for the gated deltanet recurrence and the reference implemenation of the gated deltarule. I noticed that the delta net recurrence is significantly more accurate in this case than the gated deltanet recurrence. Could this be due to some numerical issues when the gate is at 1?

![image](https://github.com/user-attachments/assets/ca5116c5-8b68-4caf-bfbe-dbbe63b1d567)

Thank you again for the great work with this library!
Julien

### Steps to reproduce the bug

```
import torch
import torch.nn.functional as F
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm
import seaborn as sns
from typing import Tuple, List, Dict


def compute_errors_between_implementations(
        T: int,
        num_trials: int = 5,
        B: int = 2,
        H: int = 2,
        D: int = 100,
        scale: float = 1.0,
        dtype: torch.dtype = torch.bfloat16,
        head_first: bool = True,
        gate_logit_normalizer: float = 1.0
) -> Dict[str, Tuple[float, float]]:
    """Compute relative errors between three implementations."""
    errors = {
        'gated_vs_ref': [],
        'delta_vs_ref': [],
        'gated_vs_delta': []
    }

    for _ in range(num_trials):
        if head_first:
            q = torch.randn(B, H, T, D, dtype=dtype)
            k = F.normalize(torch.randn(B, H, T, D, dtype=torch.float32), p=2, dim=-1).to(dtype)
            v = torch.randn(B, H, T, D, dtype=dtype)
            beta = torch.rand(B, H, T, dtype=dtype).sigmoid()
            g = torch.zeros_like(beta)
            h0 = torch.zeros(B, H, D, D, dtype=torch.float32)
        else:
            q = torch.randn(B, T, H, D, dtype=dtype)
            k = F.normalize(torch.randn(B, T, H, D, dtype=torch.float32), p=2, dim=-1).to(dtype)
            v = torch.randn(B, T, H, D, dtype=dtype)
            beta = torch.rand(B, T, H, dtype=dtype).sigmoid()
            g = torch.zeros_like(beta)
            h0 = torch.zeros(B, H, D, D, dtype=torch.float32)

        g = g / gate_logit_normalizer
        q, k, v, beta, g, h0 = map(lambda x: x.cuda().requires_grad_(True), (q, k, v, beta, g, h0))

        # Get outputs from all three implementations
        gated, gated_ht = chunk_gated_delta_rule(
            q.clone(), k.clone(), v.clone(),
            g, beta.clone(),
            scale=scale, output_final_state=True,
            initial_state=h0.clone(), head_first=head_first
        )

        delta, delta_ht = chunk_delta_rule(
            q.clone(), k.clone(), v.clone(), beta.clone(),
            scale=scale, output_final_state=True,
            initial_state=h0.clone(), head_first=head_first
        )

        ref, ref_ht = chunk_gated_delta_rule_ref(
            q.clone(), k.clone(), v.clone(), g.clone(), beta.clone(),
            scale=scale, output_final_state=True,
            initial_state=h0.clone(), head_first=head_first
        )

        # Compute relative errors
        errors['gated_vs_ref'].append((torch.norm(gated - ref) / torch.norm(ref)).item())
        errors['delta_vs_ref'].append((torch.norm(delta - ref) / torch.norm(ref)).item())
        errors['gated_vs_delta'].append((torch.norm(gated - delta) / torch.norm(delta)).item())

    # Compute statistics
    result = {}
    for comparison in errors:
        result[comparison] = (np.mean(errors[comparison]), np.std(errors[comparison]))

    return result


def analyze_implementations():
    # Sequence lengths to test
    sequence_lengths = [7, 50, 100, 200, 500, 1000]
    results = {
        'gated_vs_ref': {'means': [], 'stds': []},
        'delta_vs_ref': {'means': [], 'stds': []},
        'gated_vs_delta': {'means': [], 'stds': []}
    }

    # Print header for results table
    print("\nError Analysis Results:")
    print("-" * 85)
    print(f"{'Sequence Length':>12} | {'Implementation':>15} | {'Mean Error':>15} | {'Std Dev':>15}")
    print("-" * 85)

    # Compute errors for each sequence length
    for T in tqdm(sequence_lengths, desc="Analyzing sequence lengths"):
        comparison_results = compute_errors_between_implementations(T)

        for comparison, (mean, std) in comparison_results.items():
            results[comparison]['means'].append(mean)
            results[comparison]['stds'].append(std)

            # Print results in a formatted table
            print(f"{T:>12} | {comparison:>15} | {mean:>15.6f} | {std:>15.6f}")

    print("-" * 85)

    # Create visualization
    plt.figure(figsize=(10, 6))
    sns.set_style("whitegrid")

    colors = {'gated_vs_ref': '#2E86C1', 'delta_vs_ref': '#E74C3C', 'gated_vs_delta': '#2ECC71'}
    labels = {'gated_vs_ref': 'Gated vs Reference',
              'delta_vs_ref': 'Delta vs Reference',
              'gated_vs_delta': 'Gated vs Delta'}

    for comparison in results:
        plt.errorbar(sequence_lengths, results[comparison]['means'],
                     yerr=results[comparison]['stds'],
                     fmt='o-', capsize=5, capthick=2,
                     elinewidth=2, markersize=8,
                     color=colors[comparison],
                     label=labels[comparison])

    plt.xlabel('Sequence Length', fontsize=12)
    plt.ylabel('Relative Error', fontsize=12)
    plt.title('Relative Errors Between Implementations',
              fontsize=14, pad=20)
    plt.xscale('log')
    plt.yscale('log')
    plt.grid(True, which="both", ls="-", alpha=0.2)
    plt.legend(fontsize=10)
    plt.xticks(sequence_lengths, [str(x) for x in sequence_lengths], rotation=45)

    plt.tight_layout()
    return plt.gcf()


if __name__ == "__main__":
    # Set random seed for reproducibility
    torch.manual_seed(42)
    np.random.seed(42)

    # Run analysis
    fig = analyze_implementations()
    plt.savefig('context_length_error.pdf')
```

### Expected behavior

Error Analysis Results:
-------------------------------------------------------------------------------------
Sequence Length |  Implementation |      Mean Error |         Std Dev
-------------------------------------------------------------------------------------
           7 |    gated_vs_ref |        0.003197 |        0.000088
           7 |    delta_vs_ref |        0.003197 |        0.000088
           7 |  gated_vs_delta |        0.000000 |        0.000000
          50 |    gated_vs_ref |        0.124776 |        0.005545
          50 |    delta_vs_ref |        0.003503 |        0.000019
          50 |  gated_vs_delta |        0.125098 |        0.005510
         100 |    gated_vs_ref |        0.206634 |        0.002599
         100 |    delta_vs_ref |        0.003664 |        0.000024
         100 |  gated_vs_delta |        0.206250 |        0.002649
         200 |    gated_vs_ref |        0.232356 |        0.002785
         200 |    delta_vs_ref |        0.004023 |        0.000022
         200 |  gated_vs_delta |        0.232422 |        0.003026
         500 |    gated_vs_ref |        0.242003 |        0.001546
         500 |    delta_vs_ref |        0.004382 |        0.000016
         500 |  gated_vs_delta |        0.241797 |        0.001587
        1000 |    gated_vs_ref |        0.243902 |        0.000963
        1000 |    delta_vs_ref |        0.004520 |        0.000013
        1000 |  gated_vs_delta |        0.244336 |        0.000957
-------------------------------------------------------------------------------------

### Environment info

1. torch: 2.5.1
2. triton: 3.1

**Comment by sustcsonglin** [comment]
Hi @JulienSiems, thank you for reporting this! We speculate the issue arises from a trick we use to prevent NaN in tl.exp, and this is a corner case here. We’re working on resolving this and expect to have it fixed soon.

**Comment by yzhangcs** [comment]
```
-------------------------------------------------------------------------------------
Sequence Length |  Implementation |      Mean Error |         Std Dev
-------------------------------------------------------------------------------------
Analyzing sequence lengths:   0%|                                                                                                                                           | 0/6 [00:00<?, ?it/s]           7 |    gated_vs_ref |        0.003197 |        0.000088
           7 |    delta_vs_ref |        0.003197 |        0.000088
           7 |  gated_vs_delta |        0.000000 |        0.000000
Analyzing sequence lengths:  17%|█████████████████████▊                                                                                                             | 1/6 [00:13<01:07, 13.50s/it]          50 |    gated_vs_ref |        0.003503 |        0.000019
          50 |    delta_vs_ref |        0.003503 |        0.000019
          50 |  gated_vs_delta |        0.000000 |        0.000000
Analyzing sequence lengths:  33%|███████████████████████████████████████████▋                                                                                       | 2/6 [00:32<01:06, 16.62s/it]         100 |    gated_vs_ref |        0.003664 |        0.000023
         100 |    delta_vs_ref |        0.003664 |        0.000023
         100 |  gated_vs_delta |        0.000000 |        0.000000
Analyzing sequence lengths:  50%|█████████████████████████████████████████████████████████████████▌                                                                 | 3/6 [00:35<00:31, 10.48s/it]         200 |    gated_vs_ref |        0.004023 |        0.000022
         200 |    delta_vs_ref |        0.004023 |        0.000022
         200 |  gated_vs_delta |        0.000000 |        0.000000
Analyzing sequence lengths:  67%|███████████████████████████████████████████████████████████████████████████████████████▎                                           | 4/6 [00:38<00:15,  7.54s/it]         500 |    gated_vs_ref |        0.004382 |        0.000017
         500 |    delta_vs_ref |        0.004382 |        0.000017
         500 |  gated_vs_delta |        0.000000 |        0.000000
Analyzing sequence lengths:  83%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████▏                     | 5/        1000 |    gated_vs_ref |        0.004522 |        0.000013
        1000 |    delta_vs_ref |        0.004522 |        0.000013
        1000 |  gated_vs_delta |        0.000079 |        0.000129
Analyzing sequence lengths: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6Analyzing sequence lengths: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:44<00:00,  7.39s/it]
-------------------------------------------------------------------------------------
```

**Comment by yzhangcs** [comment]
@JulienSiems Hi, checkout https://github.com/fla-org/flash-linear-attention/commit/c57e6c2931383f6e3d16c676829710c09d399414
We've fix the masking errors. The diffs are normal now.

Thank you again for your thorough analysis!
